### PR TITLE
BUG: sparse: Fix DIA.tocoo canonical format setting

### LINKED ISSUE #10050: [Bug] inconsistent canonical format for coo_matrix
When creating an instance of `scipy.sparse.coo_matrix` from `numpy.ndarray` the constructor sets the flag `has_canonical_format` to `True` assuming that `numpy.nonzero` return indices in canonical order. But, the canonical order defined in `coo_matrix._sum_duplicates` method is `np.lexsort((row, col))` which sorts primarily by columns.

Also, running manually `sum_dumpicates()` on such matrix has no effect, even though the array is not in canonical order due to the preemptive check of `coo_matrix.has_canonical_format` flag.

The `coo_matrix.__init__` implementation:
https://github.com/scipy/scipy/blob/c3fa90dcfcaef71658744c73578e9e7d915c81e9/scipy/sparse/coo.py#L177-L187

The `coo_matrix._sum_duplicates` implentation:
https://github.com/scipy/scipy/blob/c3fa90dcfcaef71658744c73578e9e7d915c81e9/scipy/sparse/coo.py#L522-L548

### Reproducing code example:
```py
>>> import numpy as np
>>> from scipy.sparse import coo_matrix
>>> a = coo_matrix(np.arange(9).reshape(3,3))
>>> b = a.copy()
>>> b.sum_duplicates()
>>> assert a.has_canonical_format == b.has_canonical_format == True
>>> assert np.array_equal(a.row, b.row)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
AssertionError
>>> assert np.array_equal(a.col, b.col)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
AssertionError
```

### Scipy/Numpy/Python version information:
```py
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.2.1 1.16.2 sys.version_info(major=3, minor=7, micro=2, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by perimosocordiae**
#### Reference issue
Fixes gh-10050.

#### What does this implement/fix?
We were setting `has_canonical_format = True` because we knew there would be no duplicate indices, but our construction doesn't produce correctly-sorted index arrays. See the linked bug for more context.

### CODE REVIEW

**Review by dschult — APPROVED** [review]
Looks good...

File: scipy/sparse/tests/test_base.py [hunk 1]
```diff
@@ -4427,6 +4427,12 @@ def test_convert_gh14555(self):
         expected = m.toarray()
         assert_array_equal(m.tocsc().toarray(), expected)
         assert_array_equal(m.tocsr().toarray(), expected)
+    
+    def test_tocoo_gh10050(self):
+        # regression test for gh-10050
+        m = dia_matrix([[1, 2], [3, 4]]).tocoo()
+        flat_inds = np.ravel_multi_index((m.row, m.col), m.shape)
+        assert m.has_canonical_format == (np.diff(flat_inds) > 0).all()
```
Comment (jjerphan) [inline scipy/sparse/tests/test_base.py:4435]:
```suggestion
        flat_inds = np.ravel_multi_index((m.row, m.col), m.shape)
        sorted_flat_indices = (np.diff(flat_inds) > 0).all()
        assert m.has_canonical_format == sorted_flat_indices
```

**Review by jjerphan — APPROVED** [review]

File: scipy/sparse/tests/test_base.py (hunk 1 above)
Comment (perimosocordiae) [reply scipy/sparse/tests/test_base.py]: Done, thanks for the suggestion.

**Comment by jjerphan** [comment]
The failures on the CI are unrelated.

**Pull request merged**